docs(cn): translate content/blog/2020-09-22-introducing-the-new-jsx-transform.md into zh-hans (#445)
reactjs/zh-hans.reactjs.org: content/blog/2020-09-22-introducing-the-new-jsx-transform.md

---
title: "介绍全新的 JSX 转换"
author: [lunaruan]
---

虽然 React 17 [并未包含新特性](/blog/2020/08/10/react-v17-rc.html)，但它将提供一个全新版本的 JSX 转换。本文中，我们将为你描述它是什么以及如何使用。

> ## 何为 JSX 转换？ {#whats-a-jsx-transform}

Reviewer: 这些标题为啥都加了后面这一坨？

Author: 这一坨是用来做锚点的...

Author: 官方还没加，我准备去提 pr

Reviewer: 最好等他们加了咱们再加把

在浏览器中无法直接使用 JSX，所以大多数 React 开发者需依靠 Babel 或 TypeScript 来**将 JSX 代码转换为 JavaScript**。许多包含预配置的工具，例如 Create React App 或 Next.js，在其内部也引入了 JSX 转换。

> React 17 发布在即，尽管我们想对 JSX 的转换进行改进，但我们不想打破现有的配置。于是我们选择与 [Babel](https://babeljs.io/blog/2020/03/16/7.9.0#a-new-jsx-transform-11154httpsgithubcombabelbabelpull11154) 合作，为想要升级的开发者**提供了一个全新版本的，重构过的 JSX 转换**。

Reviewer: ```suggestion
React 17 发布在即，尽管我们想对 JSX 的转换进行改进，但我们不想打破现有的配置。于是我们选择[与 Babel 合作](https://babeljs.io/blog/2020/03/16/7.9.0#a-new-jsx-transform-11154httpsgithubcombabelbabelpull11154) ，为想要升级的开发者**提供了一个全新的，重构过的 JSX 转换的版本**。
```

升级至全新的转换完全是可选的，但升级它会为你带来一些好处：

* 使用全新的转换，你可以**单独使用 JSX 而无需引入 React**。
* 根据你的配置，JSX 的编译输出可能会**略微改善 bundle 的大小**。
* 它将**减少你需要学习 React 概念的数量**，以备未来之需。

**此次升级不会改变 JSX 语法，也并非必须**。旧的 JSX 转换将继续工作，没有计划取消对它的支持。

> [React 17 的 RC 版本](/blog/2020/08/10/react-v17-rc.html) 已经引入了对全新 transform 的支持，所以你可以尝试一下！为了让大家更容易使用，在 React 17 正式发布后，我们还计划将其支持 React 16.x，React 15.x 以及 React 0.14x。你可以在[下方](#how-to-upgrade-to-the-new-jsx-transform)找到不同环境的升级说明。

Reviewer: 你前面 transform 翻译成转换，这里不翻译，全文统一一个

Reviewer: ```suggestion
[React 17 的 RC 版本](/blog/2020/08/10/react-v17-rc.html) 已经引入了对全新 transform 的支持，所以你可以尝试一下！为了让大家更容易使用，在 React 17 正式发布后，我们还计划将其支持 React 16.x，React 15.x 以及 React 0.14x。你可以在[下方](#how-to-upgrade-to-the-new-jsx-transform)找到不同工具的升级说明。
```

}
```

>注意
>
>**无需改变源码**。我们将介绍 JSX 转换如何将你的 JSX 源码变成浏览器可以理解的 JavaScript 代码。

然而，这并不完美：

> * 如果使用 JSX，则需在 `React` 的环境下，因为 JSX 将被编译成 `React.createElement`。

Reviewer: ```suggestion
* 如果使用 JSX，则 `React` 需要作用域内，因为 JSX 将被编译成 `React.createElement`。
```

Author: 这里用作用域不合适吧？我觉得环境更好些

Author: 特意避开了作用域

Reviewer: 因为 scope 本身就是作用域的意思。不过问题不大，意思一样就行。早点合吧。

> * 有一些 `React.createElement` 无法做到的 [性能优化和简化](https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#motivation)。

Reviewer: ```suggestion
* 有一些 `React.createElement` 无法做到的[性能优化和简化](https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#motivation)。
```

为了解决这些问题，React 17 在 React 的 package 中引入了两个新入口，这些入口只会被 Babel 和 TypeScript 等编译器使用。新的 JSX 转换**不会将 JSX 转换为 `React.createElement`**，而是自动从 React 的 package 中引入新的入口函数并调用。例如：

```js
function App() {
  return <h1>Hello World</h1>;
}
```

现在将转换为：

```js
// 由编译器引入（禁止自己引入！）
import {jsx as _jsx} from 'react/jsx-runtime';

function App() {
  return _jsx('h1', { children: 'Hello world' });
}
```

注意，此时源代码**无需引入 React** 即可使用 JSX 了！（但仍需引入 React，以便使用 React 提供的 Hook 或其他导出。）

**此变化与所有现有 JSX 代码兼容**，所以你无需修改组件。如果你对此感兴趣，你可以查看 [RFC](https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md#detailed-design) 了解全新转换工作的具体细节。

> 注意
>
> `react/jsx-runtime` 和 `react/jsx-dev-runtime` 中的函数只能由编译器转换使用。如果你需要在代码中手动创建元素，你可以继续使用 `React.createElement`。它将继续工作，不会消失。

> ## 如何升级至新的转换 {#how-to-upgrade-to-the-new-jsx-transform}

Reviewer: ```suggestion
## 如何升级至新的 JSX 转换 {#how-to-upgrade-to-the-new-jsx-transform}
```

如果你还没准备好升级为全新的 JSX 转换，或者你正在为其他库使用 JSX，请不要担心，旧的转换不会被移除，并将继续支持。

如果你想升级，你需要准备两件事：

* **支持新转换的 React 版本**（目前，只有 [React 17 的 RC 版本](/blog/2020/08/10/react-v17-rc.html) 支持它，但是 React 17.0 发布后，我们计划针对 0.14.x、15.x 以及 16.x 做兼容。）
* **一个兼容新转换的编译器**（请看下面关于不同工具的说明）。

由于新的 JSX 转换不依赖 React 环境，[我们准备了一个自动脚本](#removing-unused-react-imports)，用于移除你代码中不必要的引入。

### Create React App {#create-react-app}

> Create React App 已[对其做兼容支持](https://github.com/facebook/create-react-app/pull/9645)，并将在 [即将发布的 v4.0 版本](https://gist.github.com/iansu/282dbe3d722bd7231fa3224c0f403fa1)中提供，该版本处于测试阶段。

Reviewer: ```suggestion
Create React App 已[对其做兼容支持](https://github.com/facebook/create-react-app/pull/9645)，并将在[即将发布的 v4.0 版本](https://gist.github.com/iansu/282dbe3d722bd7231fa3224c0f403fa1)中提供，该版本处于 beta 测试阶段。
```

}
```

### TypeScript {#typescript}

TypeScript 将在 [v4.1 beta](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1-beta/#jsx-factories) 版本中支持新的 JSX 转换。

### Flow {#flow}

Flow 将在 [v0.126.0](https://github.com/facebook/flow/releases/tag/v0.126.0) 中支持新的 JSX 转换。

## 移除未使用的 React 引入 {#removing-unused-react-imports}

因为新的 JSX 转换会自动引入必要的 `react/jsx-runtime` 函数，因此当你使用 JSX 时，将无需再引入 React。将可能会导致你代码中有未使用到的 React 引入。保留它们也无伤大雅，但如果你想删除它们，我们建议运行 [“codemod”](https://medium.com/@cpojer/effective-javascript-codemods-5a6686bb46fb) 脚本来自动删除它们：

```bash
cd your_project
npx react-codemod update-react-imports
```

>注意：
>
>如果你在运行 codemod 时出现错误，请尝试使用 `npx react-codemod update-react-imports` 选择不同的 JavaScript 环境。尤其是选择 "JavaScript with Flow" 时，即使你未使用 Flow，也可以选择它，因为它比 JavaScript 支持更新的语法。如果遇到问题，请[告知我们](https://github.com/reactjs/react-codemod/issues)。
>
>请注意，codemod 的输出可能与你的代码风格并不匹配，因此你可能需要再 codemod 完成后运行 [Prettier](https://prettier.io/) 以保证格式一致。

运行 codemod 会执行如下操作：

* 升级到新的 JSX 转换，删除所有未使用的 React 引入。

> * 改变所有 React 的默认引入将被改为解构命名引入（例如，`import React from "react"` 会变成 `import { useState } from "react"`），这将成为未来开发的首选风格。codemod **不会** 影响现有的命名空间引入方式（即 `import * as React from "react"`），这也是一种有效的风格，默认的引入将在 React 17 中继续工作，但从长远来看，我们建议尽量不使用它们。

Reviewer: ```suggestion
* 所有 React 的默认引入将被改为解构命名引入（例如，`import React from "react"` 会变成 `import { useState } from "react"`），这将成为未来开发的首选风格。codemod **不会** 影响现有的命名空间引入方式（即 `import * as React from "react"`），这也是一种有效的风格。默认的引入将在 React 17 中继续工作，但从长远来看，我们建议尽量不使用它们。
```
